instrumentation/jdbi3/src/main/java/brave/jdbi3/BraveStatementContextListener.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright The OpenZipkin Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package brave.jdbi3;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jdbi.v3.core.statement.SqlStatements;
+import org.jdbi.v3.core.statement.StatementContext;
+import org.jdbi.v3.core.statement.StatementContextListener;
+
+import brave.Span;
+import brave.Tracing;
+
+public class BraveStatementContextListener implements StatementContextListener {
+    private final Tracing tracing;
+    private static final Pattern serviceNamePattern = Pattern.compile("(^|&)zipkinServiceName=(?<serviceName>[^&]*)");
+    private Map<String, String> serviceNameCache = new HashMap<>();
```

Comment:
I suspect the `StatementContextListener` instance will be shared across many threads, should the cache be thread safe? 

```suggestion
    private Map<String, String> serviceNameCache = new ConcurrentHashMap<>();
```

Reply:
Right, it probably will be shared. Switched it to a `ConcurrentHashMap`.